test(memory): add message ordering integration tests for Issue #9909

abhiaiyer91 opened the pull request:
## Description

Add integration tests to verify message part ordering is preserved through the stream → storage → recall cycle.

## What this PR does

Adds a new test file `packages/memory/integration-tests-v5/src/message-ordering.test.ts` that:

1. **Tracks stream order** - Records TEXT, TOOL, STEP chunks as they stream from the LLM
2. **Queries raw storage** - Direct `listMessages()` query to see what's in the database
3. **Queries recall** - `memory.recall()` output to see what Memory returns
4. **Compares all three** - Verifies they match and text ordering is preserved

## Tests included

- `should preserve text ordering: stream -> raw storage -> recall`
- `should preserve ordering with multiple tool calls`
- `should match stream order exactly in storage`

## Related Issue

Closes #9909 (adds regression tests)

## How to run

```bash
cd packages/memory/integration-tests-v5
pnpm test:ordering
```

Requires `OPENAI_API_KEY` environment variable.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Tests**
  * Added new integration tests and test scripts to validate message-part ordering and streaming ID behavior across providers and storage/recall flows.

* **Bug Fixes**
  * Resolved duplicate streaming text ID issues for some providers by normalizing per-step IDs to preserve ordering and non-lossy storage/recall.

* **Chores**
  * Updated AI SDK test tooling and dev dependencies, including Anthropic and related SDK updates.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

abhiaiyer91 committed:
test(memory): add message ordering integration tests for Issue #9909

Add integration tests to verify message part ordering is preserved through
the stream -> storage -> recall cycle.

Tests compare:
- Stream order (ground truth from LLM)
- Raw storage order (direct listMessages query)
- Recall order (Memory.recall output)

These tests verify that text parts appear before tool invocations when
that was the streaming order, addressing the bug reported in Issue #9909
where text content was lost or reordered during persistence.

Related: #9909

abhiaiyer91 committed:
refactor(memory): parameterize ordering tests for OpenAI and Anthropic

abhiaiyer91 committed:
test: add integration tests for message ordering and AI SDK duplicate ID detection

- Add message-ordering.test.ts with LibSQL storage tests for both OpenAI and Anthropic
- Add ai-sdk-duplicate-ids.test.ts proving upstream bug in @ai-sdk/anthropic
- Anthropic's text-start IDs reset to 0 for each LLM call, causing duplicates in multi-step flows
- OpenAI produces unique IDs correctly

Related: #9909

abhiaiyer91 committed:
test: add Gemini to duplicate ID and message ordering tests

- Add Google Gemini to ai-sdk-duplicate-ids.test.ts to verify if it has same issue as Anthropic
- Add Gemini to message-ordering.test.ts model configs
- Tests now cover OpenAI, Anthropic, and Gemini providers

abhiaiyer91 committed:
fix: generate unique text IDs for Anthropic/Google providers

Workaround for @ai-sdk/anthropic and @ai-sdk/google duplicate text IDs bug.
These providers use numeric indices ('0', '1', etc.) that reset per LLM call,
causing duplicate text-start/text-end IDs in multi-step agentic flows.

Solution: Replace numeric IDs with UUIDs, maintaining a map per step so
text-start, text-delta, and text-end for the same block share the same UUID.

Related: #9909

abhiaiyer91 committed:
chore: add changeset for duplicate text IDs fix

DanielSLew commented on packages/core/src/stream/aisdk/v5/input.ts:
```diff
@@ -19,11 +29,40 @@ export class AISDKV5InputStream extends MastraModelInput {
+          if (!idMap.has(originalId)) {
+            idMap.set(originalId, randomUUID());
+          }
+          transformedChunk.payload.id = idMap.get(originalId)!;
```
Can't you pass in an id generator for ai sdk? Maybe we want to use that and fallback to `randomUUID()` if it's not set

abhiaiyer91 committed:
refactor: use generateId from ai-v5 instead of randomUUID

abhiaiyer91 committed:
Merge branch 'main' into olive-valley-88

abhiaiyer91 committed:
refactor: thread generateId from _internal through execute to AISDKV5InputStream

abhiaiyer91 committed:
refactor: use non-numeric IDs for text chunks in test mocks

This ensures the numeric ID workaround for Anthropic/Google doesn't
affect existing tests. Test mocks now use 'text-1' style IDs like
OpenAI does.

abhiaiyer91 committed:
fix: consistent text-delta ID in provider metadata test

abhiaiyer91 committed:
Merge branch 'main' into olive-valley-88
